### DISCUSSION TITLE
Inconsistency in OneDrive sync - a folder exists locally but not on OneDrive

### DISCUSSION THREAD

**Discussion by johnsidoli** [abraunegg/onedrive#2078] [Q&A] [answered]
I've started to use OneDrive more seriously in the past week and as part of this I had to move some files around and upload around 8,000 items in various folders.

During this exercise I also created a new folder in the route of my OneDrive (can't remember if I created it via OD web or locally in Nemo) and was playing around with some options for the folder.

Eventually I didn't want it so I deleted it in the browser OD explorer, at the same time as a lot of files were still moving around and being moved/uploaded... however this morning I noticed the test folder still existed on my local machine but not OD web.

I rebooted, and still nothing. OD was still working and would sync other items, but I still had a folder existing locally that wasn't on OD.

Two questions:

1) WHat could cause this, should I be concerned?

2) Is there a command I can run, perhaps any time whilst OD is running, just to 'check' the consistency in sync between local/OD data?

Thanks!

p.s. I am pasting some excerpts from the log around events relating to this folder, and one test doc pdf inside it:

```
2022-Aug-04 18:26:57.7610288	Received inotify delete event from operating system .. attempting item deletion as requested
2022-Aug-04 18:26:58.4800421	Creating local directory: test_as_if_this_folder
2022-Aug-04 18:26:58.7273591	Downloading file test_as_if_this_folder/test_document.pdf ... done.
...
2022-Aug-04 18:28:44.3710547	Sync with OneDrive is complete
2022-Aug-04 18:28:49.4741205	Starting a sync with OneDrive
...
2022-Aug-05 09:46:22.4891238	Received inotify delete event from operating system .. attempting item deletion as requested
2022-Aug-05 09:46:22.4898613	Deleting item from OneDrive: ./test_as_if_this_folder
2022-Aug-05 09:46:24.1589397	Starting a sync with OneDrive
2022-Aug-05 09:46:51.0980694	Sync with OneDrive is complete
2022-Aug-05 09:46:54.6734593	Uploading differences of ./test_as_if_this_folder
2022-Aug-05 09:46:54.6746497	Uploading new items of ./test_as_if_this_folder
2022-Aug-05 09:46:57.9216577	Uploading new file ./test_as_if_this_folder/test_document.pdf ... done.
2022-Aug-05 09:46:57.9243769	Starting a sync with OneDrive
2022-Aug-05 09:46:57.9244551	Syncing changes from OneDrive ...
...
2022-Aug-05 09:48:12.0144821	Sync with OneDrive is complete
2022-Aug-05 09:48:17.0960486	Starting a sync with OneDrive
2022-Aug-05 09:48:17.402103	Trying to delete item test_as_if_this_folder/test_document.pdf
2022-Aug-05 09:48:17.4034839	Deleting item test_as_if_this_folder/test_document.pdf
2022-Aug-05 09:48:17.4038358	Trying to delete item test_as_if_this_folder
2022-Aug-05 09:48:17.4048644	Deleting item test_as_if_this_folder
2022-Aug-05 09:48:17.4050978	Received inotify delete event from operating system .. attempting item deletion as requested
2022-Aug-05 09:48:17.4054759	Received inotify delete event from operating system .. attempting item deletion as requested
```

**Comment by abraunegg** [comment, marked as answer]
@johnsidoli 
> WHat could cause this, should I be concerned?

Microsoft only send 1 event for actions performed online .. and, if for whatever reason, it is not processed or received locally - because of some other action / activity - Microsoft never send it again.

It is weird that it got lost to begin with - but without knowing the exact sequence of events to try an replicate I cant advise it was this or that which caused it.

> Is there a command I can run, perhaps any time whilst OD is running, just to 'check' the consistency in sync between local/OD data?

You can use the `--display-sync-status` command - but that implies no active running client. You can add `--dry-run` to the mix, but you will not then have the most up-to-date data locally and the client will report that a few items are still out-of-sync.

You can also use the `--resync` to force the client & online to get back-in-sync .. but that is more of a brute-force method.

The last option would be to use webhooks to subscribe to remote updates in 'monitor' mode - this way, when an online change occurs, it is sent in a separate channel and processed differently. See: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#use-webhook-to-subscribe-to-remote-updates-in-monitor-mode for more information.

**Comment by abraunegg** [comment]
@johnsidoli 
> ```
>2022-Aug-05 09:48:17.402103	Trying to delete item test_as_if_this_folder/test_document.pdf
>2022-Aug-05 09:48:17.4034839	Deleting item test_as_if_this_folder/test_document.pdf
>2022-Aug-05 09:48:17.4038358	Trying to delete item test_as_if_this_folder
>2022-Aug-05 09:48:17.4048644	Deleting item test_as_if_this_folder
>2022-Aug-05 09:48:17.4050978	Received inotify delete event from operating system .. attempting item deletion as requested
>2022-Aug-05 09:48:17.4054759	Received inotify delete event from operating system .. attempting item deletion as requested
>```

OK .. from this it reads as though the local database ID for the folder or item is not equal to what is online ... perhaps from changes / moves done ... 

The best option to fix this would be:
1. stop the client, stop any systemd service
2. remove the offending local folder
3. restart the client manually adding `--resync`
4. once completed, restart any systemd service

I would also look at setting up webhooks as well.

**Reply by johnsidoli** [reply]
> The best option to fix this would be:
> 
>     1. stop the client, stop any systemd service
> 
>     2. remove the offending local folder
> 
>     3. restart the client manually adding `--resync`
> 
>     4. once completed, restart any systemd service

Just to note, for the slower ones amongst us, a cheatsheet of these commands would be incredibly helpful.

To begin it would be this right?

Stop systemd service: systemctl stop onedrive@\<username>.service
Restart client: onedrive --synchronize --resync
Restart systemd: systemctl start onedrive@\<username>.service

Also, is this a good way to 'resync' and check things for sharepoint too? I am so paranoid all the time about getting any of this wrong and things going awry.

**Comment by johnsidoli** [comment]
Thanks @abraunegg ... I did already delete the folder and everything **seems** fine now.

I will read into the webhooks but to be honest this might be a step too far for me at the moment, in technical terms. I am not an advanced user so getting this far has been a mission haha.
